src/utils/array.js

```diff
@@ -202,47 +202,75 @@ function _resize (array, size, dim, defaultValue) {
  */
 export function reshape (array, sizes) {
   const flatArray = flatten(array)
-  let newArray
-
-  function product (arr) {
-    return arr.reduce((prev, curr) => prev * curr)
-  }
+  const currentLength = flatArray.length
 
   if (!Array.isArray(array) || !Array.isArray(sizes)) {
     throw new TypeError('Array expected')
   }
 
   if (sizes.length === 0) {
-    throw new DimensionError(0, product(arraySize(array)), '!=')
-  }
-
-  let totalSize = 1
-  for (let sizeIndex = 0; sizeIndex < sizes.length; sizeIndex++) {
-    totalSize *= sizes[sizeIndex]
+    throw new DimensionError(0, currentLength, '!=')
   }
 
-  if (flatArray.length !== totalSize) {
+  sizes = processSizesWildcard(sizes, currentLength)
+  const newLength = product(sizes)
+  if (currentLength !== newLength) {
     throw new DimensionError(
-      product(sizes),
-      product(arraySize(array)),
+      newLength,
+      currentLength,
       '!='
     )
   }
 
   try {
-    newArray = _reshape(flatArray, sizes)
+    return _reshape(flatArray, sizes)
   } catch (e) {
     if (e instanceof DimensionError) {
       throw new DimensionError(
-        product(sizes),
-        product(arraySize(array)),
+        newLength,
+        currentLength,
         '!='
       )
     }
     throw e
   }
 
-  return newArray
+  function product (arr) { return arr.reduce((prev, curr) => prev * curr) }
+}
+
+/**
+ * Replaces the wildcard -1 in the sizes array.
+ * @param {Array.<number>} sizes  List of sizes for each dimension. At most on wildcard.
+ * @param {number} currentLength  Number of elements in the array.
+ * @throws {Error}                If more than one wildcard or unable to replace it.
+ * @returns {Array.<number>}      The sizes array with wildcard replaced.
+ */
+export function processSizesWildcard (sizes, currentLength) {
+  const newLength = product(sizes)
+  const processedSizes = sizes.slice()
+  const WILDCARD = -1
+  const wildCardIndex = sizes.indexOf(WILDCARD)
+  ensureAtMostOneWildcard()
+
+  if (hasWildcard()) {
+    if (canReplaceWildcard()) {
+      replaceWildcard()
+    } else {
+      throw new Error('Could not replace wildcard, since ' + currentLength + ' is no multiple of ' + (-newLength))
+    }
+  }
+  return processedSizes
+
+  function ensureAtMostOneWildcard () {
+    const isMoreThanOneWildcard = sizes.indexOf(WILDCARD, wildCardIndex + 1) >= 0
+    if (isMoreThanOneWildcard) {
+      throw new Error('More than one wildcard in sizes')
+    }
+  }
+  function hasWildcard () { return wildCardIndex >= 0 }
+  function canReplaceWildcard () { return currentLength % newLength === 0 }
+  function replaceWildcard () { processedSizes[wildCardIndex] = -currentLength / newLength }
+  function product (arr) { return arr.reduce((prev, curr) => prev * curr) }
```

Comment:
Is this style ok? I wanted to keep the logic short.

Reply:
I understand it is nice to have less lines of code, but I think for readability it's important to just use add a couple of new lines.

Looking at the code itself however, I think you can lift `prod` outside of `processSizesWildcard`, and the other functions can be simply changed into constants and/or inlined?

Reply:
I wanted to keep the main logic as clean as possible by extracting the functions (and reduce bloat by not using three lines per function).

Regarding inlining `replaceWildcard()`: I just prefer calling the function this way instead of inlining it and adding a comment `// replace wildcard`. But if you prefer the style of inlining, I'm totally fine with it. 

And I didn't want to lift out `prod`, since it gets used only there in this file. 

Would you prefer consts and comments?

Reply:
I understand. Extracting a piece of code in a separate function can make the code easier to follow and understand. But it also comes at a cost: you introduce an extra indirection and extra code. Replacing one line of code which is used in one occurrence only with a function is a bit on the edge I think. In this case I think both approaches would be ok though I personally prefer with the inline constants solution like you did in the end.